packages/cozy-stack-client/src/OAuthClient.js

```diff
@@ -492,37 +494,63 @@ class OAuthClient extends CozyStackClient {
    * @returns {Promise} A promise that resolves with a new AccessToken object
    */
   async refreshToken() {
-    if (!this.isRegistered()) throw new NotRegisteredException()
-    if (!this.token) throw new Error('No token to refresh')
+    if (!this.isRegistered()) {
+      throw new NotRegisteredException()
+    }
+    if (!this.token) {
+      throw new Error('No token to refresh')
+    }
 
     const data = {
       grant_type: 'refresh_token',
       refresh_token: this.token.refreshToken,
       client_id: this.oauthOptions.clientID,
       client_secret: this.oauthOptions.clientSecret
     }
+    try {
+      const result = await this.fetchJSONWithCurrentToken(
+        'POST',
+        '/auth/access_token',
+        this.dataToQueryString(data),
+        {
+          headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
+        }
+      )
 
-    const result = await super.fetchJSON(
-      'POST',
-      '/auth/access_token',
-      this.dataToQueryString(data),
-      {
-        headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
+      const newToken = new AccessToken({
+        refresh_token: this.token.refreshToken,
+        ...result
+      })
+
+      this.setToken(newToken)
+
+      if (this.onTokenRefresh && typeof this.onTokenRefresh === 'function') {
+        this.onTokenRefresh(newToken)
       }
-    )
 
-    const newToken = new AccessToken({
-      refresh_token: this.token.refreshToken,
-      ...result
-    })
+      return newToken
+    } catch (e) {
+      if (errors.UNREGISTERED_CLIENT.test(e.message)) {
+        this.onRevocationChange(true)
 
-    this.setToken(newToken)
+        throw new Error(
+          'Client has been revoked or does not exist. Please authenticate again.'
+        )
```

Comment:
I don't know if we should change the error message. Shouldn't we throw the real error? Or a fetch error?

Reply:
Well yes I guess I wanted to imitate the checkForRevocation conditional
```js
        const revoked = await this.checkForRevocation()

        if (revoked) {
          throw new Error('Client has been revoked. Please authenticate again.')
        }
    ```
    
 I'm not sure if it's still a fetch error at this point, it's more a reaction to a fetch error. But we could indeed at least keep the error message and add more info saying that error message probably happens because client was either revoked or doesn't exist (both can be true)